iris-mpc-cpu/src/execution/hawk_main/state_check.rs

```diff
@@ -0,0 +1,112 @@
+use eyre::{eyre, Result};
+use itertools::{chain, Itertools};
+use serde::{Deserialize, Serialize};
+use siphasher::sip::SipHasher13;
+use std::hash::{Hash, Hasher};
+
+use super::{HawkSession, HawkSessionRef};
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+pub struct SetHash {
+    accumulator: u64,
+}
+
+impl SetHash {
+    pub fn add_unordered(&mut self, value: impl Hash) {
+        self.accumulator = self.accumulator.wrapping_add(Self::hash(value));
+    }
+
+    pub fn remove(&mut self, value: impl Hash) {
+        self.accumulator = self.accumulator.wrapping_sub(Self::hash(value));
+    }
```

Comment:
Think the accumulation here is fine for the use-case but should keep in mind that this method + the data types here allow a pretty easy ~2^32 cost collision attack.

Reply:
Definitely not cryptographic hash, it is a checksum.

Reply:
Renamed to `checksum`.